src/main/scala/fi/oph/koski/validation/KoskiValidator.scala

```diff
@@ -691,4 +692,26 @@ class KoskiValidator(tutkintoRepository: TutkintoRepository, val koodistoPalvelu
       case _ => HttpStatus.ok
     }
   }
+
+  private def validateOsaamisenHankkimistavat(suoritus: Suoritus): HttpStatus = {
```

Comment:
Tätä saisi vielä tiivistettyä esim

```suggestion
  private def validateOsaamisenHankkimistavat(suoritus: Suoritus): HttpStatus = suoritus match {
    case a: AmmatillisenTutkinnonSuoritus =>
      HttpStatus.validate(
        a.osaamisenHankkimistavat.toList.flatten
          .map(_.osaamisenHankkimistapa)
          .collect { case o: OsaamisenHankkimistapaIlmanLisätietoja => o }
          .forall(hankkimistapa => List("koulutussopimus", "oppilaitosmuotoinenkoulutus").contains(hankkimistapa.tunniste.koodiarvo))
      )(KoskiErrorCategory.badRequest.validation.rakenne("foobar"))
    case _ => HttpStatus.ok
  }
```
- optinaalisten listojen käsittely on ärsyttävää, mutta se voidaan ohittaa muuttamalla optionaalinen lista pelkäksi listaksi `.toList.flatten` kikalla. jos optinaalinen lista olisi tyhjä (`None`) niin tällöin `.toList.flatten`in jälkeen palautuisi `Nil` mikä scalassa tarkoittaa `List` tyyppin kohdalla tyhjää listaa.
- `HttpStatus.scala`sta löytyy käteviä funktiota validaatioiden tekemiseen
- ehdottaisin kanssa, että käytettäisiin koodiarvojen validaatiossa whitelistausta, jotta uusien koodiarvojen lisäys ei aiheuta muutoksia mistä ei olla varmoja. uusien whitelistien lisääminen on pienempi paha vrt. mahdollisesti epävalidi data

Reply:
Tää onkin ovela. Selkeä, ja `forall` tosiaan short-circuittaa myös. 💡